### PR TITLE
 DAOS-623 build: Various build improvements

### PR DESCRIPTION
**Pull request by brianjmurrell**
Add get_base_branch utility to walk the commit history backwards until
a base branch is found.

Update packaging:
- use mock's ccache plugin.
- use new get_base_branch utility from the packaging update to add the
base branch to the environment in BASE_BRANCH_NAME for other stages in
the pipeilne to be able to use.

Use pragmasToEnv() rather than a bigger script that does the same thing
to conserve Jenkinsfile pipeline {} size.

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
  * Hardware testing was skipped as it adds no value to testing the changes
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: utils/scripts/get_base_branch
Comment (ashleypittman) [inline utils/scripts/get_base_branch:1]: FYI - the new shellcheck checker relies on shell scripts have a .sh extension.

Comment (brianjmurrell) [reply utils/scripts/get_base_branch:1]: Hrm.  That's less than optimal.  Typically [shell] scripts that you want to be regularly executable commands have a shebang at the top of them and don't typically have a [`.sh`] extension on them and scripts you want to be _libraries_ or otherwise executed with a specific interpreter don't have a shebang but do have a [`.sh`] extension on the end.

I.e.:
https://google.github.io/styleguide/shellguide.html#s2.1-file-extensions
https://www.talisman.org/~erlkonig/documents/commandname-extensions-considered-harmful/
as just a couple of references to the practice.

As an alternative `file` could be used to determine which linter to use on a file.  But I think you told me that we lint the entire code base now, on every commit and not just the files in the commit, yes?  I would suppose running `file` on every file in the tree, on every commit might be onerous.

```
$ time bash -c 'git ls-files | xargs file >/dev/null'

real    0m6.215s
user    0m6.007s
sys     0m0.130s
```

Comment (brianjmurrell) [reply utils/scripts/get_base_branch:1]: So yeah.  If they are not executable and don't have a shebang, those should have a `.sh` extension.  If we fixed that, we'd be all good then it sound like, yes?

Comment (ashleypittman) [reply utils/scripts/get_base_branch:1]: Then we'd want CI to check for all files with a .sh extension *and* all files that "file" identifies as a shell script?  That would be ideal yes, but I don't think it's what we do today.

I think most of the discussion was in https://github.com/daos-stack/daos/pull/13360 and I remember being surprised that we were missing files with the old system.

Comment (ashleypittman) [reply utils/scripts/get_base_branch:1]: On a clean checkout "find . -type f -exec file '{}' \; | grep "shell script" | grep -v "\.sh"" shows 18 commands, three of which are in vendor and two are .in files which are renamed in https://github.com/daos-stack/daos/pull/13076 and should not be checked as they're not shell.

**Review by JohnMalmberg — DISMISSED** [review]

File: utils/rpms/packaging/get_base_branch
Comment (daltonbohning) [inline utils/rpms/packaging/get_base_branch:16]: When we branch release/2.6 from master, during the short time that master == release/2.6, before we merge new commits into master, this will break because it will print two branches like
```
master
release/2.6
```

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:338]: Ha.  I had forgotten about env.CHANGE_TARGET.  Although the argument that could be made for `get_base_branch` is that it's independent of infrastructure like GitHub or GitLab, etc and works with `git` alone.  I'm not sure that's an important consideration at this point though.

File: utils/rpms/packaging/get_base_branch
Comment (brianjmurrell) [reply utils/rpms/packaging/get_base_branch]: Good catch.  I haven't pushed it yet but I have a different algorithm for determining this more along the lines of the one used in the git hooks.

File: utils/rpms/packaging/get_base_branch
Comment (daltonbohning) [inline utils/rpms/packaging/get_base_branch:23]: As we discussed, checking `commits_behind` like `find_base.sh` does is only for edge cases (if even useful at all). Could we either (A) update `find_base.sh` to be in-sync or (B) share the function?

Also, is `utils/rpms/packaging` the right place for this, if only called from the Jenkinsfile?

Comment (brianjmurrell) [reply utils/rpms/packaging/get_base_branch:23]: > As we discussed, checking `commits_behind` like `find_base.sh` does is only for edge cases (if even useful at all).

Right, which is why I omitted that in this implementation.

> Could we either (A) update `find_base.sh` to be in-sync or (B) share the function?

My original intent was to share/genericize the function but you seemed to want to keep the `commits_behind` functionality in `find_base.sh` and I didn't really see the need for it here.  Indeed, it nags at me that it might make finding the true base (where we disregard intermediate branches such as `features/*`) actually less accurate.

> Also, is `utils/rpms/packaging` the right place for this, if only called from the Jenkinsfile?

It seems odd, yes, but this functionality is used across projects (`daos` and all of the dependency packaging projects as it's used in the dependency packaging `pipeline`s) and `packaging/` is the common denominator there -- without inventing yet another cross-project library.

Comment (daltonbohning) [reply utils/rpms/packaging/get_base_branch:23]: FWIW, I'd rather give up `commits_behind` in `find_base.sh` than have two slightly different versions.

Comment (daltonbohning) [reply utils/rpms/packaging/get_base_branch:23]: > this functionality is used across projects

It is currently, or it will be after landed?

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:338]: Actually, I take that back, and am going to have to update this PR to reflect that.  `env.CHANGE_TARGET` is not useful in this case as it could be any arbitrary branch.  Any arbitrary branch doesn't tell us the ultimate landing branch which is what we are looking for here.  Using the _base branch_ nomenclature in this stage is probably misleading since GH uses that nomenclature to describe what `env.CHANGE_BRANCH` is -- the merge target of a change -- but that's not what we are looking for technically.

File: Jenkinsfile
Comment (daltonbohning) [inline Jenkinsfile:332]: "Landing" branch might be misleading too. If someone opens a PR to be merged into "foo", then "foo" is the immediate landing branch.

Comment (brianjmurrell) [reply Jenkinsfile:332]: Where `foo` is based on what?  Just so that we are using the same nomenclature, the branch against which a PR is create is called the _base branch_.  I'm specifically not looking for a base branch here.  For example, if somebody opens a PR on a branch called `foo` with a base branch of `master` and then opens a PR on a branch named `bar` against the branch `foo`, `foo` is the _base branch_ but `master` is the _landing branch_.  It's that latter, _landing branch_ name that I am looking for.

File: utils/rpms/packaging/get_base_branch
Comment (brianjmurrell) [reply utils/rpms/packaging/get_base_branch]: There are a few moving parts that are working towards this, but it will be used by all project pipelines that build RPMs (hence the association with `packaging/`.

Comment (daltonbohning) [reply utils/rpms/packaging/get_base_branch]: I do think it's better for this function to be in `daos` instead of `pipeline-lib` - for re-use purposes as well as ease-of-updating.

The question of "where do I put a script that is used for CI, hooks, etc." has come up before, and I'm not sure where the best place is to put them. I don't want to block on this, so I'm okay with what you have and we can revisit later if/when we start re-using this function elsewhere

File: Jenkinsfile
Comment (daltonbohning) [reply Jenkinsfile]: Right. The devil is in the nuance :) I think "release" might be the most appropriate.

**Review by daltonbohning — DISMISSED** [review]

**Comment by brianjmurrell** [comment]
This one is ready for [re-]review please and thanks, in advance.

File: Jenkinsfile
Comment (daltonbohning) [inline Jenkinsfile:332]: Mostly a reminder for myself that when I get the auto-test-tag magic working, we might need to rethink this a bit - or I rethink my approach

File: Jenkinsfile
Comment (daltonbohning) [inline Jenkinsfile:326]: What is "environment block"?

File: Jenkinsfile
Comment (daltonbohning) [inline Jenkinsfile:336]: How does `releaseBranch` call `get_release_branch`?

File: utils/githooks/find_base.sh
Comment (daltonbohning) [inline utils/githooks/find_base.sh:39]: I'm almost tempted to say we don't need to look at the feature branches, since as you pointed out before, they should be up-to-date with master anyway.

But I think we could handle that later after consulting with other relevant parties

**Review by daltonbohning — APPROVED** [review]

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:332]: I don't think this should have any impact.  This is just an optimization to store the commit pragmas into an env. variable rather than everywhere wanting to examine them running a `git` command.  Ultimately `cachedCommitPragma()` reads them from the env. var. if it exists and falls back to a `git` command if not.

And this use of `pragmasToEnv()` isn't really a change in behaviour.  It's just a more code efficient way of what was being done before.

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:326]: https://github.com/daos-stack/daos/blob/bead33473c2aaa9b93645f0122021188df4f6a14/Jenkinsfile#L141-L150

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:336]: Heh.  https://github.com/daos-stack/daos/blob/bead33473c2aaa9b93645f0122021188df4f6a14/Jenkinsfile#L141-L150

File: Jenkinsfile
Comment (daltonbohning) [reply Jenkinsfile:332]: In my current version, I replace
```
sh(script: 'git show -s --format=%B',
                                            returnStdout: true).trim()
```
With a call to `get_commit_message.py`.

So now the question is how to call `get_commit_message.py` before `pragmasToEnv()`. But we can look at that later. The work is still some ways off

File: Jenkinsfile
Comment (daltonbohning) [reply Jenkinsfile:336]: Sorry if double posted, but it seems GH lost some of my comments...

This
`daos:/Jenkinsfile`
calls
`pipeline-lib:/vars/releaseBranch.groovy`
which calls
`daos:/utils/rpms/packaging/get_release_branch`

IMO that's messy to bounce back and forth between repos

File: utils/githooks/find_base.sh
Comment (brianjmurrell) [reply utils/githooks/find_base.sh:39]: It is.  We actually have lots of that going on because it's a necessary evil we need to do to minimize the amount of code we add to `Jenkinsfile`'s `pipeline {}` block due to the design flaw of how Jenkins renders that into Java.

Comment (daltonbohning) [reply utils/githooks/find_base.sh:39]: From my limited understanding of the issue, it's a limit on the _**static**_ pipeline size. But function calls should by dynamic. So I would think something like this should work, even putting groovy scripts in the `daos` repo
```
load("my_func.groovy").foo()
```

Anyway, that's a much larger problem to address so doesn't need to block this PR.

**Review by JohnMalmberg — APPROVED** [review]

**Pull request merged**
